apps/dpnk/fakturoid_invoice_gen.py

```diff
@@ -0,0 +1,143 @@
+
+    :param class instance invoice: Invoice model class instance
+
+    :return list lines: list of InvoiceLine class model instaces
+    """
+
+    lines = []
+    for payment in invoice.payment_set.order_by(
+        "user_attendance__userprofile__user__last_name",
+        "user_attendance__userprofile__user__first_name",
+    ):
+        if invoice.company_pais_benefitial_fee:
+            amount = invoice.campaign.benefitial_admission_fee_company
+        else:
+            amount = payment.amount
+        user_name = (
+            "" if invoice.anonymize else payment.user_attendance.name_for_trusted()
+        )
+        description = f"Platba za soutěžící/ho {user_name}"
+        lines.append(
+            # https://fakturoid.docs.apiary.io/#reference/lines
+            InvoiceLine(
+                name=description,
+                quantity=1,
+                unit_price=decimal.Decimal(amount),
+                vat_rate=21,
+            )
+        )
+    return lines
+
+
+def create_or_update_invoice(fa, subject, invoice):
+    """Create or update Fakturoid invoice
+
+    :param class instance fa: fakturoid.Fakturoid class instance
+    :param class instance subject: fakturoid.Subject model class instance
+    :param class instance: Invoice model class instance
+    """
+    # Invoice
+    fa_invoice = fa.invoices(custom_id=invoice.id)
+    try:
+        fa_invoice = list(fa_invoice)[0]
+    except IndexError:
+        fa_invoice = None
+    fa_invoice_data = {
+        "custom_id": invoice.id,
+        "subject_id": subject.id,
+        "number": f"{invoice.exposure_date.year}-{invoice.sequence_number}",
+        "order_number": invoice.sequence_number,
+        "lines": [],
+    }
+    if not fa_invoice:
+        fa_invoice_data["lines"] = create_invoice_lines(invoice=invoice)
+        fa_invoice = Invoice(**fa_invoice_data)
+    else:
+        fa_invoice.lines = []
+        fa.save(fa_invoice)
+        fa_invoice_data["lines"] = create_invoice_lines(invoice=invoice)
+        fa_invoice.update(fa_invoice_data)
+    fa.save(fa_invoice)
+
+
+def get_fakturoid_api(account):
+    """Get Fakturoid API class instance
+
+    :param str account: Fakturoid account type "production" or "test"
+
+    :return class instance fakturoid.Fakturoid: Fakturoid API class
+                                                instance
+    """
+    return Fakturoid(
+        slug=settings.FAKTUROID[account]["user_account"],
+        email=settings.FAKTUROID[account]["user_email"],
+        api_key=settings.FAKTUROID[account]["api_key"],
+        user_agent=settings.FAKTUROID[account]["user_agent_header"],
+    )
+
+
+def generate_invoice(invoice, fakturoid_account=None):
+    """Generate Fakturoid invoice
+
+    :param class instance: Invoice model class instance
+    :param str account: Fakturoid account type "production" or "test"
+    """
+    if not fakturoid_account:
+        if invoice.created < settings.FAKTUROID["date_from_create_invoices"]:
```

Comment:
Tohle nerozumím...

Reply:
Pre prepojenie fakturoidu s DPNK, potrebujeme 2 Fakturoid accounts **test** a **production**. 

Premenná `settings.FAKTUROID["date_from_create_invoices"]` znamená že **od tohto dátumu sa budú generovať faktúry vo Fakturoide,** zároveň bude vytvorený nový model Invoice v DPNK, ale admin Invoice model už nebude editovateľný. Viď https://github.com/auto-mat/do-prace-na-kole/pull/233/commits/44172f883ff1a8cddedb085f23898e2cf511b7ad#diff-ade6d60b8264300a632d72cba1d2db95374d23c31eedfa1b16c5a4b10c56c7c1R1649. Ondřej chce aby sa to Fakturoide faktúrovalo až od budúceho ročníka DPNK.

Do dátumu v premennej `settings.FAKTUROID["date_from_create_invoices"]` budeme, pre testovanie používať testovací Fakturoid account **test** na testovacej instancií DPNK.

Reply:
No, ale tuhle funkce se nevola do dátumu `settings.FAKTUROID["date_from_create_invoices"]` ani na testu. A pokud chceš mít jiné nastavení na testovací instancií DPNK, tak jde prostě nastavit jiné env hodnoty pro testovaci instance.

Reply:
To čo checem dosiahnúť aby po nasadení na testovaciu a produkčnú instanciu DPNK sa používal najprv **testovací** Fakturoid account do dátumu nastaveného v globálnej premennej `FAKTUROID_DATE_FROM_CREATE_INVOICES`, a následne ak bude všetko v poriadku od tohto dátumu sa bude používať **prudukčný** Fakturoid account pre **reálne** faktúrovanie, tak to odporúčajú vo fakturoide.

Pre unit testy sa bude naďalej používať **test** Fakturoid account.